backup: panic: runtime error: invalid memory address or nil pointer dereference

sqlsmith was recently taught about backups. While running one in roachtest, this seemed to cause a panic:

https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_WorkloadNightly/1676715:id/sqlsmith/setup%3Drand-tables/setting%3Ddefault/1.logs/cockroach.log

```
E200107 07:57:33.001786 2252 storage/cloud/external_storage.go:570  [n1,client=35.243.196.42:58210,user=root] HTTP:Req error: err=retryable http error: Get http://127.0.0.1:34055/backup_1/BACKUP: dial tcp 127.0.0.1:34055: connect: connection refused (attempt 29)
E200107 07:57:35.041402 2252 storage/cloud/external_storage.go:570  [n1,client=35.243.196.42:58210,user=root] HTTP:Req error: err=retryable http error: Get http://127.0.0.1:34055/backup_1/BACKUP: dial tcp 127.0.0.1:34055: connect: connection refused (attempt 30)
E200107 07:57:37.084856 2252 storage/cloud/external_storage.go:570  [n1,client=35.243.196.42:58210,user=root] HTTP:Req error: err=retryable http error: Get http://127.0.0.1:34055/backup_1/BACKUP: dial tcp 127.0.0.1:34055: connect: connection refused (attempt 31)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x1d66e01]

goroutine 2252 [running]:
panic(0x3c15c40, 0x6ac1cb0)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc0056428f8 sp=0xc005642868 pc=0x790a32
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:199
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:394 +0x3ec fp=0xc005642928 sp=0xc0056428f8 pc=0x7a664c
github.com/cockroachdb/cockroach/pkg/storage/cloud.newResumingHTTPReader(0x492b920, 0xc006dbefc0, 0xc0067e6f90, 0x410ed68, 0xf, 0x0, 0x48af600, 0xc000231200)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/cloud/external_storage.go:520 +0xb1 fp=0xc005642970 sp=0xc005642928 pc=0x1d66e01
github.com/cockroachdb/cockroach/pkg/storage/cloud.(*httpStorage).ReadFile(0xc0067e6f90, 0x492b920, 0xc006dbefc0, 0x410ed68, 0xf, 0x48eb520, 0x0, 0x48af600, 0xc000231200)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/cloud/external_storage.go:649 +0x57 fp=0xc0056429c0 sp=0xc005642970 pc=0x1d67eb7
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.VerifyUsableExportTarget(0x492b920, 0xc006dbefc0, 0xc00019ca80, 0x496e860, 0xc0067e6f90, 0xc006476da0, 0x1f, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup.go:945 +0xd7 fp=0xc005642a60 sp=0xc0056429c0 pc=0x2cd3be7
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.backupPlanHook.func1(0x492b920, 0xc006dbefc0, 0x0, 0x0, 0x0, 0xc005f90420, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup.go:1260 +0x187b fp=0xc005643ef0 sp=0xc005642a60 pc=0x2cf778b
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1(0xc006a21380, 0x492b920, 0xc006dbefc0, 0xc005685440, 0xc006f30710)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:145 +0x7d fp=0xc005643fb8 sp=0xc005643ef0 pc=0x297bacd
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc005643fc0 sp=0xc005643fb8 pc=0x7c10a1
created by github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:144 +0xcc
```

Queries at https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_WorkloadNightly/1676715:id/sqlsmith/setup%3Drand-tables/setting%3Ddefault/sqlsmith.log

--- a/pkg/storage/cloud/external_storage.go
+++ b/pkg/storage/cloud/external_storage.go
@@ -559,6 +559,10 @@ func checkHTTPContentRangeHeader(h string, pos int64) error {
 func (r *resumingHTTPReader) sendRequest(
 	reqHeaders map[string]string,
 ) (resp *http.Response, err error) {
+	// Initialize err to the context.Canceled: if our context is canceled, we will
+	// never enter the loop below; in this case we want to return "nil, canceled"
+	err = context.Canceled
+
 	for attempt, retries := 0,
 		retry.StartWithCtx(r.ctx, httpRetryOptions); retries.Next(); attempt++ {
 		resp, err = r.client.req(r.ctx, "GET", r.url, nil, reqHeaders)
@@ -570,7 +574,7 @@ func (r *resumingHTTPReader) sendRequest(
 		log.Errorf(r.ctx, "HTTP:Req error: err=%s (attempt %d)", err, attempt)
 
 		if _, ok := err.(*retryableHTTPError); !ok {
-			break
+			return
 		}
 	}
 

--- a/pkg/storage/cloud/external_storage_test.go
+++ b/pkg/storage/cloud/external_storage_test.go
@@ -878,3 +878,22 @@ func TestHttpGet(t *testing.T) {
 		})
 	}
 }
+
+func TestHttpGetWithCancelledContext(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	defer s.Close()
+
+	store, err := makeHTTPStorage(s.URL, testSettings)
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, store.Close())
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	_, err = store.ReadFile(ctx, "/something")
+	require.Error(t, context.Canceled, err)
+}

